Initial implementation of model inputs

crisbeto opened the pull request:
Adds support for model inputs in the framework. `model()` returns a writable signal that implicitly defines a input/output pair that can be used either in two-way bindings to keep two values in sync or by binding individually to the input and output. When the value of the `model` changes, it will emit an event with the current value.

Furthermore, these changes expand two-way bindings to accept `WritableSignal`. This will make it easier to transition existing code to signals in a backwards-compatible way.

Example:

```ts
@Directive({
  selector: 'counter',
  standalone: true,
  host: {
    '(click)': 'increment()',
  }
})
export class Counter {
  value = model(0);

  increment(): void {
    this.value.update(current => current + 1);
  }
}

@Component({
  template: `<counter [(value)]="count"/> The current count is: {{count()}}`,
})
class App {
  count = signal(0);
}
```

**Note:** with these changes model inputs should be mostly working. There are a few places where things will be improved in a follow-up PR:
- Ensuring that the language service works correctly with the new TCB for two-way bindings.
- Avoid having to return an object literal from `ModelSignal.subscribe`.
- Figure out the performance implications of checking if the value is a signal each time the `twoWayProperty` instruction is called.
- Figure out if we should remove `applyValueToInputSignal` from `MODEL_SIGNAL_NODE`.
- Avoid the extra variable in the two-way binding TCB.
- Adding a diagnostic when `input()` is mixed with `@Output()` in a two-way binding.

crisbeto committed:
refactor(core): add model implementation

Adds the implementation of the `model` primitive that represents a two-way binding signal-based binding.

crisbeto committed:
refactor(compiler-cli): add function to identify model fields

Adds a function to the compiler to help us identify fields initialized to a `model`.

crisbeto committed:
refactor(compiler-cli): add jit transform for model inputs

Adds a JIT transform that marks `model` fields as `@Input` and `@Output`.

crisbeto committed:
refactor(core): implement two-way instructions

Adds the implementations of the `twoWayProperty` and `twoWayListener` instructions.

crisbeto committed:
refactor(compiler-cli): recognize model fields

Adds the logic that recognizes fields initialized to model as an input/output pair.

crisbeto committed:
test(core): add runtime acceptance tests for model inputs

Sets up the runtime tests for model inputs.

crisbeto committed:
refactor(compiler-cli): allow writable signals in two-way bindings

Updates the TCB generation logic to allow for `WritableSignal` to be assigned in two-way bindings.

crisbeto committed:
refactor(compiler-cli): add tests for model inputs

Adds tests in the compiler to verify the compiled output and template type checking behavior of model inputs.

kbrilla commented:
Question: in example we have `model(0)` and `singal(0)` so both have same starting value, but which one will win in scenario `model(5)` and `signal(0)` ?

crisbeto commented:
`model(0)` sets the default value and it'll be overwritten by the one coming from the property binding.

crisbeto committed:
feat(core): add support for model inputs

Adds support for model inputs in the framework. `model()` returns a writable signal that implicitly defines a input/output pair that can be used either in two-way bindings to keep two values in sync or by binding individually to the input and output. When the value of the `model` changes, it will emit an event with the current value.

Furthermore, these changes expand two-way bindings to accept `WritableSignal`. This will make it easier to transition existing code to signals in a backwards-compatible way.

Example:

```ts
@Directive({
  selector: 'counter',
  standalone: true,
  host: {
    '(click)': 'increment()',
  }
})
export class Counter {
  value = model(0);

  increment(): void {
    this.value.update(current => current + 1);
  }
}

@Component({
  template: `<counter [(value)]="count"/> The current count is: {{count()}}`,
})
class App {
  count = signal(0);
}
```

crisbeto committed:
refactor(compiler-cli): account for model inputs in language service

Updates the code that resolves a node from the TCB to an input in order to fix the language service.

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
````diff
@@ -846,6 +848,52 @@ class TcbDirectiveInputsOp extends TcbOp {
+    // following expressions to expand the type:
+    // ```
+    // var captureType = dir.value;
+    // (id as unknown as ɵConditionallyUnwrapSignal<typeof captureType> |
````
Context here is that I had to keep the expression in the form of `dir.value = expression`, because the language service looks for assignment expression specifically. The `as unknown` can technically be omitted, but it improves the diagnostic message in some edge cases.

devversion replied:
👍 yeah, this was a key design principle for signal inputs as well, and I brought that up for model too

JeanMeche commented on packages/core/src/errors.ts:
```diff
@@ -118,6 +118,7 @@ export const enum RuntimeErrorCode {
 
   // Signal inputs
   REQUIRED_INPUT_NO_VALUE = -950,
+  REQUIRED_MODEL_NO_VALUE = -951,
```
You might already want to take 952 instead of 951. See #54103

JeanMeche commented on packages/core/src/core.ts:
```diff
@@ -13,10 +13,11 @@
 
 export * from './authoring';
-// Input is exported separately as this file is exempted from JSCompiler's
+// Input and Mode are exported separately as these files are exempted from JSCompiler's
```
```suggestion
// Input and Model are exported separately as these files are exempted from JSCompiler's
```

JeanMeche commented on packages/core/src/authoring/model/model_signal_node.ts:
```diff
@@ -0,0 +1,37 @@
+
+import {SIGNAL_NODE, SignalNode, signalSetFn} from '@angular/core/primitives/signals';
+
+export const REQUIRED_UNSET_VALUE = /* @__PURE__ */ Symbol('ModelSignalNode#UNSET');
```
What is the purpose on having different `UNSET` symbols for each type of signal ? (Input/computed/Query)

devversion commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
````diff
@@ -846,6 +848,52 @@ class TcbDirectiveInputsOp extends TcbOp {
+    // allows `val` to be `number | string | WritableSignal<number | string>`. We generate the
+    // following expressions to expand the type:
+    // ```
+    // var captureType = dir.value;
````
A maybe more readable variant could be:

```ts
var _t1: (ɵConditionallyUnwrapSignal<X> | WritableSignal<ɵConditionallyUnwrapSignal<X>>) = null!:
(dir1.field as typeof _t1) = dir2.otherMatchingInput = <user-expr>
```

that would keep the assignment clearer I think. Or does this also yield the worse diagnostic?

devversion commented on packages/compiler-cli/src/ngtsc/testing/fake_core/index.ts:
```diff
@@ -136,6 +136,8 @@ export function signal<T>(initialValue: T): WritableSignal<T> {
   return null!;
 }
 
+export type ɵConditionallyUnwrapSignal<T> = T extends Signal<unknown>? ReturnType<T>: T;
```
Should we only bother about writable signals? as this is what model implements? I think this would make sure we don't accidentally unwrap two way bindings to fields taking a `Signal`.

alxhub replied:
+1, we explicitly require writable signals for two-way bindings.

devversion commented on packages/compiler-cli/test/ngtsc/authoring_spec.ts:
```diff
@@ -199,6 +199,33 @@ runInEachFileSystem(() => {
           .toContain(`inputs: { data: [i0.ɵɵInputFlags.SignalBased, "data"] }`);
     });
 
+    // TODO(crisbeto): we may not want to support this combination. Will discuss with the team.
```
Good conversation. I believe we want to support use-cases where the local "state" might be desychronized from the parent. There is a good issue about people asking for this- although in practice we couldn't come up with a good scenario where users _may not_ want to properly synchronize parent & component (e.g. `Dir#expanded`)

alxhub replied:
This combination is fine - there's an input and output of the right shape for two-way binding. We can't _enforce_ that components conform to the two-way binding contract.

devversion commented on packages/compiler-cli/test/ngtsc/authoring_spec.ts:
```diff
@@ -284,4 +311,391 @@ runInEachFileSystem(() => {
+          '{ "value": "valueChange"; }, never, never, false, never>;');
+    });
+
+    it('should declare an input/output pair for a field initialized to an alised model()', () => {
```
Nit: `aliased`

devversion commented on packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts:
```diff
@@ -350,15 +350,32 @@ export class SymbolBuilder {
+          ts.isAsExpression(node.left.expression.expression) &&
+          isAccessExpression(node.left.expression.expression.expression)) {
+        assignment = node.left.expression.expression.expression;
+      }
```
Maybe already planned as a follow-up: Can you please add a test for the language service. Definition + completions? I invested a lot in writing tests here and I think we should continue doing this.

devversion commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -846,6 +848,52 @@ class TcbDirectiveInputsOp extends TcbOp {
+    // the time of writing, this generates a suboptimal diagnostic message where TS splits up the
+    // signature, e.g. "Type 'number' is not assignable to type 'string | boolean |
+    // WritableSignal<string> | WritableSignal<false> | WritableSignal<true>'" instead of Type
+    // 'number' is not assignable to type 'string | boolean | WritableSignal<string | boolean>'.
```
Can you add the language service symbol builder reason here as well?

devversion reviewed (approved):
LGTM overall, mostly nits/questions. Nice! 🎉

crisbeto committed:
refactor(compiler-cli): rework TCB for two-way bindings

Reworks the TCB for two-way bindings to make them simpler and to avoid regressions for two-way bindings to generic inputs. The new TCB looks as follows:

```
var _t1: Dir;
var _t2 = _t1.input;
(_t1 as typeof _t2 | WritableSignal<typeof _t2>) = expression;
```

crisbeto commented on packages/core/src/render3/instructions/two_way.ts:
```diff
@@ -27,8 +32,21 @@ import {ɵɵproperty} from './property';
+    value = value() as T;
+  }
+
+  const lView = getLView();
```
I was asking about this in one of our channels. It's because instructions aren't supposed to call directly into other instructions (even though there are a handful of places that do it).

crisbeto commented on packages/core/test/acceptance/authoring/model_inputs_spec.ts:
```diff
@@ -0,0 +1,546 @@
+
+    // Changing the value from within the directive.
+    host.dir.value.set(2);
+    fixture.detectChanges();
```
Not necessary, removing.

crisbeto commented on packages/core/test/acceptance/authoring/model_inputs_spec.ts:
```diff
@@ -0,0 +1,546 @@
+    expect(host.dir.value()).toBe(3);
+  });
+
+  it('should allow a two-way-bound signal to be bound in the template', () => {
```
I wanted to capture that the change are reflected in the DOM. Will update the description.

crisbeto commented on packages/core/test/acceptance/authoring/model_inputs_spec.ts:
```diff
@@ -0,0 +1,546 @@
+    class Dir implements OnChanges {
+      value = model(0);
+
+      ngOnChanges(allCahnges: SimpleChanges): void {
```
I think there's a test somewhere with `effect` already. I put this test here mostly to ensure that things work correctly because `ngOnChanges` stores previous values.

crisbeto committed:
refactor(core): address PR feedback

Addresses the feedback from #54252.

crisbeto commented:
I've reworked the TCB to avoid some regressions. Also addressed all the feedback in the last commit.

devversion commented on packages/compiler-cli/test/ngtsc/authoring_spec.ts:
```diff
@@ -696,6 +727,120 @@ runInEachFileSystem(() => {
+          standalone: true,
+        })
+        export class TestDir<T> {
+          value = model.required<T>();
```
We should also have a test to ensure the `value` property is inferred properly. This is a follow-up I guess?

devversion commented on packages/core/src/render3/instructions/two_way.ts:
```diff
@@ -27,8 +32,21 @@ import {ɵɵproperty} from './property';
-  // TODO(crisbeto): implement two-way specific logic.
-  ɵɵproperty(propName, value, sanitizer);
+  // TODO(crisbeto): perf impact of re-evaluating this on each change detection?
+  if (isSignal(value)) {
```
Should this be `isWritableSignal`?

crisbeto replied:
No since we also allow `input` signal to passed directly into a `model` when it's done through the template.

devversion replied:
Why wouldn't the user use a normal input binding then?

crisbeto replied:
They can't unwrap the `input` value inside a two-way binding so it has to be passed by reference.

alxhub replied:
I'm almost certain this should be `isWritableSignal`. You shouldn't be able to two-way bind to an `input()` - inputs are readonly.

devversion replied:
They would use an `[x]="myInput()"` binding. They are clearly not interested in the output event

crisbeto replied:
Changed.

atscott reviewed (approved):
reviewed-for: public-api

alxhub commented on packages/core/src/authoring/model/model_signal.ts:
```diff
@@ -0,0 +1,117 @@
+    }
+  };
+
+  getter.update = (updateFn: (value: T|typeof REQUIRED_UNSET_VALUE) => T) => {
```
Does this `typeof` need to be here? We'll never call the user's function with it afaik. Calling `.update` when the signal isn't set in the first place should 💥

crisbeto replied:
It is, otherwise TS produces the following:

> Argument of type 'T | unique symbol' is not assignable to parameter of type 'T'.

alxhub replied:
It's clearly wrong as a type though, which suggests that something is possible that shouldn't be. We should never be calling the user's `update` function with `REQUIRED_UNSET_VALUE`.

If the user calls `.update()` on an unset `model()`, it should throw the same error as if they tried to read it. We should have a test for this.

crisbeto replied:
Fixed.

crisbeto committed:
refactor(compiler-cli): fix regression in two-way bindings to inputs with different getter/setter types

In a previous commit the TCB was changed to cast the assignment to an input in order to widen its type to allow `WritableSignal`. This ended up breaking existing inputs whose setter has a wider type than its getter. These changes switch to unwrapping the value on the binding side.

crisbeto committed:
refactor(core): assert writable signal in two-way property instruction

Asserts that the value is a `WritableSignal`, rather than a `Signal`, in the `twoWayProperty` instruction.

crisbeto commented on packages/core/src/render3/instructions/two_way.ts:
```diff
@@ -27,8 +32,21 @@ import {ɵɵproperty} from './property';
  */
 export function ɵɵtwoWayProperty<T>(
     propName: string, value: T, sanitizer?: SanitizerFn|null): typeof ɵɵtwoWayProperty {
```
Changed in the last commit.

alxhub reviewed (approved):
Reviewed-for: public-api

LGTM modulo the one comment about `update()` throwing an error

crisbeto committed:
refactor(core): throw if required model is changed via update too early

Adds an assertion that will throw if `ModelSignal.update` is accessed before an initial value is set.

crisbeto committed:
refactor(core): correctly distinguish getter functions from writable signals

Fixes that `ɵunwrapWritableSignal` inferring getter functions as not matching the interface of `WritableSignal` instead of preserving them.

crisbeto commented:
[Passing TGP](http://test/OCL:604605717:BASE:604930003:1707306132966:b576445e)

jessicajaniuk commented:
This PR was merged into the repository by commit e921e108e1837781b168ae16aafdc56a64607afc.

sysmat commented:
17.2.0-rc.0 model not working as classical prop

```html
<select [(ngModel)]="selected">
      <option [value]="7">7</option> <-- init value is not selected
      <option [value]="14">14</option>
</select>
{{ selected() }} <-- init value is not display

...

selected = model(7);
```

JeanMeche commented:
`ngModel` is a directive, it does not use the new `model()`  input. (Which isn't ready in itself also).

crisbeto commented:
@JeanMeche `model()` is ready to be used. There are some internal cleanups to do, but they won't affect the behavior.

wartab commented:
I know things aren't completed yet, but is there anything planned at framework level for forms? Or is the following method the way to go for the foreseeable future?

```html
<input [ngModel]="signal()" (ngModelChange)="signal.set($event)">
```

sysmat commented:
@wartab yes as classical prop as signal, for now I'm doing it this way

sysmat commented:
@JeanMeche so is there another directive for model?

crisbeto commented:
@sysmat I just tried your example with the RC and it seems to work. See https://stackblitz.com/edit/stackblitz-starters-dcjjnw?file=src%2Fmain.ts

@wartab two-way binding to a signal is supported with this change. You should be able to do `[(ngModel)]="someSignal"`.

wartab commented:
Really cool, thank you very much

sysmat commented:
@crisbeto yes it works with **signal**, but not with **model** is a bit of mystery to me why not with **model**, probably I'm missing something here

crisbeto commented:
It works for me with `model()` as well. Here's a fork of the previous commit: https://stackblitz.com/edit/stackblitz-starters-kij9th?file=src%2Fmain.ts. Maybe you haven't updated the compiler to the latest RC?

sysmat commented:
- I have
```
"@angular/compiler-cli": "^17.2.0-rc.0",
"@angular/compiler": "^17.2.0-rc.0",
```

- @crisbeto super, yes I don't now why in my local setup dosen't work

crisbeto commented:
It may be worth clearing your `node_modules` and reinstalling.

sysmat commented:
@crisbeto I delete lock, re install, and also using     **"builder": "@angular-devkit/build-angular:application",**

sysmat commented:
@crisbeto in your stackblitz you are using, is this ok

```
"devDependencies": {
    "@angular-devkit/build-angular": "~0.1100.4",
    "@angular/cli": "~11.0.4",
    "@angular/compiler-cli": "~11.0.4",
    "@types/jasmine": "~3.6.0",
    "@types/node": "^12.11.1",
    "codelyzer": "^6.0.0",
    "jasmine-core": "~3.6.0",
    "jasmine-spec-reporter": "~5.0.0",
    "karma": "~5.1.0",
    "karma-chrome-launcher": "~3.1.0",
    "karma-coverage": "~2.0.3",
    "karma-jasmine": "~4.0.0",
    "karma-jasmine-html-reporter": "^1.5.0",
    "protractor": "~7.0.0",
    "ts-node": "~8.3.0",
    "tslint": "~6.1.0",
    "typescript": "~4.0.2"
```

sysmat commented:
my component is loaded by router, but this shouldn't be problem, when it works with signal

```
{
    path: 'model',
    loadComponent: async () => (await import('./app/ng-model.component')).SelectModelSignalComponent,
  }
```

crisbeto commented:
Hmm that's weird. Here's a fork with 17.2 of the CLI: https://stackblitz.com/edit/stackblitz-starters-9pipk1?file=package.json

crisbeto commented:
Following up here: we've released version 17.2.0-rc.1 to resolve the type checking error that @JeanMeche flagged above.

sysmat commented:
- when creating fresh project with cli, it works
```
Angular CLI: 17.2.0-rc.0
Node: 20.11.0
Package Manager: npm 10.2.4
OS: win32 x64

Angular:
...

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.1702.0-rc.0 (cli-only)
@angular-devkit/core         17.2.0-rc.0 (cli-only)
@angular-devkit/schematics   17.2.0-rc.0 (cli-only)
@schematics/angular          17.2.0-rc.0 (cli-only)
```

```
"dependencies": {
    "@angular/animations": "^17.2.0-next.0",
    "@angular/common": "^17.2.0-next.0",
    "@angular/compiler": "^17.2.0-next.0",
    "@angular/core": "^17.2.0-next.0",
    "@angular/forms": "^17.2.0-next.0",
    "@angular/platform-browser": "^17.2.0-next.0",
    "@angular/platform-browser-dynamic": "^17.2.0-next.0",
    "@angular/router": "^17.2.0-next.0",
```

- VERSION.full --> 17.2.0-rc.1

- model & select babnana in the box work

```html
<select [(ngModel)]="selectedModel">
        <option [value]="7">7</option>
        <option [value]="14">14</option>
</select>
<hr />
Selected value: {{ selectedModel() }}

....
version = VERSION.full;
selectedModel = model(7);
```

- so with my old project started as ng 9, doing ng update and migrations, is not working with model but it works with signal

sysmat commented:
- in my new project when using routing then not working **model & [()]**

```
export const routes: Routes = [
    {
        path: 'model',
        loadComponent: async () => (await import('./ng-model.component')).SelectModelSignalComponent,
      },
];
```

```
export const appConfig: ApplicationConfig = {
  providers: [
    provideRouter(routes, 
    withComponentInputBinding(),
    withViewTransitions(),
    withInMemoryScrolling({
      scrollPositionRestoration: 'enabled',
    }))]
};
```

sysmat commented:
I take fork of @crisbeto with routing and it is not working on route (model.ts), but work in main

- https://stackblitz.com/edit/stackblitz-starters-vjekma?file=src%2Fmain.ts
- https://stackblitz-starters-vjekma.stackblitz.io

crisbeto commented:
I'm guessing this is an issue when setting the value through `setInput`, which is used by the router.

sysmat commented:
yes, the **withComponentInputBinding()** is the problem

scottwalter-nice commented:
> yes, the **withComponentInputBinding()** is the problem

 If I have a component that has no inputs but have withComponentInputBinding() enabled.   Declaring a model such as 
 modelName = model('Model Scott'); will not appear in the template:  {{ modelName()}}
 
 Unless I set the value in ngInit:    modelName.set('Model Scott');

sysmat commented:
@scottwalter-nice but why should me **model** bound to router, for that we have **input**

scottfwalter commented:
> @scottwalter-nice but why should me **model** bound to router, for that we have **input**

I was probably misusing the shiny new model when input or a simple Signal would have sufficed.  I guess we need a good explainer on basic SIgnal vs model.

crisbeto commented:
@sysmat looking into your Stackblitz: I think the problem is that you aren't setting a value for the `selectedModel` parameter so the router sets it to `undefined`. E.g. if I set the link to `[routerLink]="['/model', {selectedModel: 7}]"` it works fine. Here's an updated Stackblitz https://stackblitz.com/edit/stackblitz-starters-gjjmie.

scottwalter-nice commented:
> @sysmat looking into your Stackblitz: I think the problem is that you aren't setting a value for the `selectedModel` parameter so the router sets it to `undefined`. E.g. if I set the link to `[routerLink]="['/model', {selectedModel: 7}]"` it works fine. Here's an updated Stackblitz https://stackblitz.com/edit/stackblitz-starters-gjjmie.

To be honest I’m confused between signal inputs and model.  I originally thought model was for two way binding,  but i can do that with regular signals on 17.2

What is the use case of model vs signal inputs?

crisbeto commented:
1. `model()` is a shorthand to define a two-way binding.
2. `model` signals are writable from inside the component while the `input()` signals are a read-only.

imaksp commented:
> > @scottwalter-nice but why should me **model** bound to router, for that we have **input**
> 
> I was probably misusing the shiny new model when input or a simple Signal would have sufficed. I guess we need a good explainer on basic SIgnal vs model.

yes I also used new model for ngModel field initially but it didn't work with 'withComponentInputBinding',  should have tried this normal signal first. I assume its not meant for using with [(ngModel)].
